src/main/java/com/epam/aidial/core/service/ApplicationService.java

```diff
@@ -279,4 +489,420 @@ private void prepareApplication(ResourceDescription resource, EtagHeader etag,
+                },
+                (response, body) -> convertServerSentEvent(body, CreateDeploymentResponse.class));
+
+        return deployment.url();
+    }
+
+    // endregion
+
+    // region Terminate Application
+
+    private Void terminateApplication(ResourceDescription resource, String error) {
+        try (LockService.Lock lock = lockService.tryLock(deploymentLockKey(resource))) {
+            if (lock == null) {
+                return null;
+            }
+
+            Application application;
+
+            try {
+                application = getApplication(resource).getValue();
+            } catch (ResourceNotFoundException e) {
+                application = null;
+            }
+
+            if (isPending(application)) {
+                Application.Function function = application.getFunction();
+
+                if (!isPublicOrReview(resource)) {
+                    deleteFolder(function.getTargetFolder());
+                }
+
+                deleteApplicationImage(function);
+                deleteApplicationDeployment(function);
+
+                resourceService.computeResource(resource, json -> {
+                    Application existing = ProxyUtil.convertToObject(json, Application.class);
+                    if (existing == null || !Objects.equals(existing.getFunction(), function)) {
+                        throw new IllegalStateException("Application function has been updated");
+                    }
+
+                    Application.Function.Status status = (function.getStatus() == Application.Function.Status.STOPPING)
+                            ? Application.Function.Status.STOPPED
+                            : Application.Function.Status.FAILED;
+
+                    function.setStatus(status);
+                    function.setError(status == Application.Function.Status.FAILED ? error : null);
+
+                    existing.setFunction(function);
+                    return ProxyUtil.convertToString(existing);
+                });
+            }
+
+            pendingApplications.remove(resource.getUrl());
+            return null;
+        } catch (Throwable e) {
+            log.warn("Failed to terminate application: {}", resource.getUrl(), e);
+            throw e;
+        }
+    }
+
+    private void deleteApplicationImage(Application.Function function) {
+        callController(HttpMethod.DELETE, "/v1/image/" + function.getId(),
+                request -> null,
+                (response, body) -> convertServerSentEvent(body, EmptyResponse.class));
+    }
+
+    private void deleteApplicationDeployment(Application.Function function) {
+        callController(HttpMethod.DELETE, "/v1/deployment/" + function.getId(),
+                request -> null,
+                (response, body) -> convertServerSentEvent(body, EmptyResponse.class));
+    }
+
+    // endregion
+
+    private void copyFolder(String sourceFolderUrl, String targetFolderUrl) {
```

Comment:
Let's move the method to ResourceService

Reply:
done